build/v1/types.go

```diff
@@ -977,7 +977,11 @@ type BuildConfigStatus struct {
 	LastVersion int64 `json:"lastVersion" protobuf:"varint,1,opt,name=lastVersion"`
 
 	// ImageChangeTriggers is used to capture the runtime state of any ImageChangeTrigger specified in the BuildConfigSpec,
-	// including reconciled values of the lastTriggeredImageID and paused fields.
+	// including the value reconciled by the OpenShift APIServer for the lastTriggeredImageID.  There will be single entry
```

Comment:
this is suggesting that the openshift-apiserver is setting a status field?  That is pretty unusual.  What causes the openshift-apiserver to set this?

Reply:
> this is suggesting that the openshift-apiserver is setting a status field? That is pretty unusual. What causes the openshift-apiserver to set this?

if the openshift-apiserver is setting this value, can you describe which resource or subresource you're hitting that forces this calculation?

Reply:
iirc, a buildrequest comes into the apiserver.  The apiserver processes that buildrequest against the buildconfig to create a new build object.  As part of that, it must reconcile the current SHA values.

Once the apiserver has created the build object, it updates the buildconfig w/ information about the most recent build object it created.

Why buildrequests are fully processed in the apiserver(assuming my memory is correct about how it works) as part of the request flow and not in a controller is probably an artifact of the time period in which the code was written.  It'd probably be a controller loop today.

Reply:
@bparees 's account of the flow, including the fact the original time period on when the feature was introduced influencing the implementation then, and how things have been left untouched since then, are what I discovered while doing this feature.

I've certainly applied "minimal change needed" to my mindset while working on this change.  But of course if we reach a consensus that we want to take on additional changes to further "modernize", I'll react accordingly.  I have no strong opinion wrt that.

I'll put some code links and historical PR refs for reference.  I'll use existing merged code vs. what I have in https://github.com/openshift/openshift-apiserver/pull/191

1) ICTs for Builds was introduced in Aug 2015 :-) where the setting of the `LastTriggeredImageID` field in the apiserver first occurred at https://github.com/openshift/origin/blob/603eab741040fbb06219de02587ca6dce45837b2/pkg/build/generator/generator.go#L226 .... if you look at the `updateImageTriggers` back then, it does not look too different from today at https://github.com/openshift/openshift-apiserver/blob/7c32a64c6f1ea0d7a3e8f96bce0232d48c568eac/pkg/build/apiserver/buildgenerator/generator.go#L342-L382

2) Similarly, the flow that leads to the call to `updateImageTrigger` have remained similar.  Back in Aug 2015 it occurred in the `Instantiate` method.  See https://github.com/openshift/origin/blob/603eab741040fbb06219de02587ca6dce45837b2/pkg/build/generator/generator.go#L168
Then in Instantiate, the client `UpdateBuildConfig` was called.

The same occurs today, see https://github.com/openshift/openshift-apiserver/blob/7c32a64c6f1ea0d7a3e8f96bce0232d48c568eac/pkg/build/apiserver/buildgenerator/generator.go#L251 from `instantiate` for the call to `updateImageTriggers`, followed by the call the the client `UpdateBuildConfig` at https://github.com/openshift/openshift-apiserver/blob/7c32a64c6f1ea0d7a3e8f96bce0232d48c568eac/pkg/build/apiserver/buildgenerator/generator.go#L316

3) that `UpdateBuildConfig` in today's code ultimately translates to this REST invocation:  https://github.com/openshift/openshift-apiserver/blob/f85871147bb7c03c544fbc9c6515048ec470844d/vendor/github.com/openshift/client-go/build/clientset/versioned/typed/build/v1/buildconfig.go#L112-L123